### DISCUSSION TITLE
Extension API for GitHub copilot

### DISCUSSION THREAD

**Discussion by Susanta120** [microsoft/vscode-discussions#1101] [Extension Development QnA]
Hi there, 
I am trying to get text that I have written in GitHub copilot chat in VS code. But not found any extension API. Can anyone help me here? Or any other way to get this?
Thanks in advance.

**Comment by abrilurena** [comment]
Hi, @Susanta120! I hope you're doing great. To use GitHub Copilot Chat in Visual Studio Code, follow these steps:

1. Install the GitHub Copilot Chat extension:
- Open Visual Studio Code.
- Click on the Extensions icon in the activity bar on the left side of the window.
- In the search box “Search Extensions in Marketplace,” look for “GitHub Copilot Chat” and click Install.
- After installation, click Reload Required to update Visual Studio Code.
2. Start using GitHub Copilot Chat:
- You should now be able to use GitHub Copilot Chat in Visual Studio Code.
- You can access Copilot Chat in different ways:
- Chat View: Ask Copilot in the chat view for help with any task or question, and it will provide code suggestions based on your current files.
- Inline Chat: Apply Copilot’s AI suggestions directly to your code without leaving your workflow.

Remember that to use GitHub Copilot Chat, you need an active GitHub Copilot subscription. If you have any further questions or need more assistance, feel free to ask! 😊

**Reply by Susanta120** [reply]
Thanks @abrilurena for your explanation. But I want the chat text that I have written in copilot inline chat box. And I want to get this using vscode extension api. Is there any way?

**Reply by Susanta120** [reply]
Hi there, 
Can I detect keyboard event using vscode api? Like if I press ctrl+v, can I listen this event?

**Comment by tamuratak** [comment]
You would be interested in

- https://code.visualstudio.com/api/extension-guides/chat

> Visual Studio Code's [Copilot Chat](https://marketplace.visualstudio.com/items?itemName=GitHub.copilot-chat) architecture enables extension authors to integrate with the Copilot Chat experience. A chat extension is a VS Code extension that uses the Chat extension API by contributing a Chat participant.

- https://github.com/microsoft/vscode-extension-samples/tree/main/chat-sample

> This sample shows
>
>   - How to contribute a chat participant.
>   - How to use the chatRequestAccess API to request access to the chat.
>   - How to respond with follow-ups.

- microsoft/vscode/issues/199908

**Reply by Susanta120** [reply]
Thanks @tamuratak , I will go through this.

**Reply by Susanta120** [reply]
Hi there, 
I am facing a problem that when I delay the save event, it's not delaying, it gets save immediately. I want to delay the document save event. So that I can perform some logic.
Here is my code:
```
let saveDisposal = vscode.workspace.onWillSaveTextDocument((event) => {
//event.preventDefault();
console.log("Save event call------");
event.waitUntil(func);
	});
async function func () {
		return new Promise ((resolve, reject) => {
			setTimeout(() => {
				console.log("110----Saved");
				resolve(true);
			}, 10000)
		})
	}
```
When I am saving saving document **Save event call------** get printed. But **110----Saved** is not get printed.
Can anyone help me?

**Comment by theReynald** [comment]
Hi, @Susanta120! 
Hmmmm. In a VS Code extension, I don't think you can directly delay the save event using **event.waitUntil()**. The **onWillSaveTextDocument** event is not asynchronous, and calling **event.waitUntil()** won't defer the save operation.

However, you can try using a different approach. You can listen for the **onWillSaveTextDocument** event to intercept the save action and perform your logic asynchronously. Once your logic is complete, you can then trigger the save operation manually.

Here's how you can modify your code to achieve this:

```JavaScript
let saveDisposal = vscode.workspace.onWillSaveTextDocument(async (event) => {
    console.log("Save event call------");
    // Perform your asynchronous logic
    await func();
    // After the logic is done, trigger the save operation manually
    vscode.window.showTextDocument(event.document);
});

async function func() {
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            console.log("110----Saved");
            resolve(true);
        }, 10000);
    });
}

```

In this modified code:

We listen for the **onWillSaveTextDocument** event.
When the event is triggered, we log a message indicating that the save event has been called.
We then asynchronously perform your logic using the func() function.
After the logic is completed (in your case, after a delay of 10 seconds), we trigger the save operation manually using **vscode.window.showTextDocument(event.document)**.

This way, your logic will be executed before the document is saved.

**Reply by Susanta120** [reply]
Thanks @theReynald , it worked.
I have another question: I am using proxy server to intercept requests from vscode. I need vscode user email and username in request header who is currently logged in. How can I get this?
